includes/Core/Key_Metrics/Key_Metrics_Settings.php

```diff
@@ -49,6 +49,28 @@ protected function get_default() {
 		return array();
 	}
 
+	/**
+	 * Merges an array of settings to update. Only existing keys will be updated.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param array $partial Partial settings array to save.
+	 * @return bool True on success, false on failure.
+	 */
+	public function merge( array $partial ) {
+		$settings = $this->get();
+
+		if ( isset( $partial['isWidgetHidden'] ) && is_bool( $partial['isWidgetHidden'] ) ) {
+			$settings['isWidgetHidden'] = $partial['isWidgetHidden'];
+		}
+
+		if ( ! empty( $partial['widgetSlugs'] ) && is_array( $partial['widgetSlugs'] ) ) {
+			$settings['widgetSlugs'] = $partial['widgetSlugs'];
+		}
```

Comment:
Wouldn't it be better to add `is_bool` and `is_array` checks to the conditions in `get_sanitize_callback`, rather than here?

Reply:
It will be too late. The idea here is that if some properties don't match expected type, then we don't merge it into settings.

Reply:
Hmm, OK. It seems a bit odd though, to have a check in `merge` to ensure the exact type is passed in, but then have code in `get_sanitize_callback`, to massage the data into the correct format if the exact type is _not_ passed in. Why not just pass the data straight through and let `get_sanitize_callback` take care of it? It will ensure that no incorrectly formed data will be stored and it would be easier to manage keeping all data checks of this sort within the one method...

Reply:
@techanvil, ok let's assume I did what you asked and moved sanitization to the `get_sanitize_callback` callback. Let's see how it works now. Assuming we have the following settings:
```php
array(
    'widgetSlugs'    => array( 'moduleA', 'moduleB' ),
    'isWidgetHidden' => false,
)
```
Then the `merge` function is called with wrong data:
```php
$key_metrics_settings->merge( array(
    'widgetSlugs'    => null,
    'isWidgetHidden' => true,
) );
```
Since we don't validate the incoming data in the `merge` function, we have wrong data merged and after the sanitization callback is executed, we will get the `widgetSlugs` array reset to an empty array because `null` is invalid value - as a result we get data loss.

In other words instead:
```php
array(
    'widgetSlugs'    => array( 'moduleA', 'moduleB' ),
    'isWidgetHidden' => true,
)
```
We will get:
```php
array(
    'widgetSlugs'    => array(),
    'isWidgetHidden' => false,
)
```

> It seems a bit odd though, to have a check in `merge` ...

There is nothing odd there. The `get_sanitize_callback` callback is responsible for data sanitization before saving it into the database. The sanitization in the `merge` function is responsible for the incoming data verification before changing the current settings.

Reply:
> However, it raises the question, why should `merge` behave differently to `set`?

Because it allows partial updates.

> If we move the check for `is_array` etc into `get_sanitize_callback` then we can have consistent behaviour for both `merge` and `set`

The `get_sanitize_callback` callback already takes care of the `widgetsSlug` property by calling the `Sanitize:: sanitize_string_list` method. I don't think we should change anything there. At least not in this ticket, because it is out of scope for this one.

https://github.com/google/site-kit-wp/blob/8825f41e392caa787480ae033c8eed5d0e4e174a/includes/Core/Util/Sanitize.php#L30-L47

Reply:
What I meant was, why should the validation/sanitization for `merge` behave differently to `set`? It just seems a bit odd to be have, say, `merge( array( 'widgetSlugs' => 'foo' ) )` result in the settings _not_ being updated, while the same arguments to `set` results in the data being updated (with `foo` wrapped in an array).

Conceptually I think `merge` should just be like `set` except it only sets a slice of the data.

Sorry, I don't mean to be obtuse or excessively pedantic here, it's a pretty small detail but I do think it's a good opportunity to avoid introducing an inconsistent interface before it's actually used anywhere. I do take your point that it's arguably out of scope to modify `get_sanitize_callback`, but in that case why not just pass the data through without checking the type? It will still be sanitized by `get_sanitize_callback` which can be our single source of truth for data sanitization.

Also as you've mentioned elsewhere, by adding the API schema we have an extra layer of validation that will enforce these data types for the API calls.

Reply:
> but in that case why not just pass the data through without checking the type? 

Sounds good, I have reverted my changes and copied the `merge` method from the `Module_Settings` class as IB suggests.